src/libsync/bulkpropagatorjob.cpp

```diff
@@ -0,0 +1,869 @@
+{
+    QByteArray ocEtag = OCC::parseEtag(reply.value("OC-ETag").toString().toLatin1());
+    QByteArray etag = OCC::parseEtag(reply.value("ETag").toString().toLatin1());
+    QByteArray ret = ocEtag;
+    if (ret.isEmpty()) {
+        ret = etag;
+    }
+    if (ocEtag.length() > 0 && ocEtag != etag) {
+        qCDebug(OCC::lcBulkPropagatorJob) << "Quite peculiar, we have an etag != OC-Etag [no problem!]" << etag << ocEtag;
+    }
+    return ret;
+}
+
+inline QByteArray getHeaderFromJsonReply(const QJsonObject &reply, const QByteArray &headerName)
+{
+    return reply.value(headerName).toString().toLatin1();
+}
+
+}
+
+namespace OCC {
+
+BulkPropagatorJob::BulkPropagatorJob(OwncloudPropagator *propagator,
+                                     const std::deque<SyncFileItemPtr> &items)
+    : PropagatorJob(propagator)
+    , _items(items)
+{
+    _uploadFileParameters.reserve(_items.size());
+}
+
+bool BulkPropagatorJob::scheduleSelfOrChild()
+{
+    if (_items.empty()) {
+        return false;
+    }
+
+    _state = Running;
+    for(int i = 0; i < 100 && !_items.empty(); ++i) {
+        auto currentItem = _items.front();
+        _items.pop_front();
+        QMetaObject::invokeMethod(this, [this, currentItem] () {
+            UploadFileInfo fileToUpload;
+            fileToUpload._file = currentItem->_file;
+            fileToUpload._size = currentItem->_size;
+            fileToUpload._path = propagator()->fullLocalPath(fileToUpload._file);
+            startUploadFile(currentItem, fileToUpload);
+        }); // We could be in a different thread (neon jobs)
+    }
+
+    return _items.empty() && _uploadFileParameters.empty();
+}
+
+PropagatorJob::JobParallelism BulkPropagatorJob::parallelism()
+{
+    return PropagatorJob::JobParallelism::WaitForFinished;
+}
+
+void BulkPropagatorJob::startUploadFile(SyncFileItemPtr item, UploadFileInfo fileToUpload)
+{
+    if (propagator()->_abortRequested) {
+        return;
+    }
+
+    // Check if the specific file can be accessed
+    if (propagator()->hasCaseClashAccessibilityProblem(fileToUpload._file)) {
+        done(item, SyncFileItem::NormalError, tr("File %1 cannot be uploaded because another file with the same name, differing only in case, exists").arg(QDir::toNativeSeparators(item->_file)));
+        return;
+    }
+
+    // Check if we believe that the upload will fail due to remote quota limits
+    const qint64 quotaGuess = propagator()->_folderQuota.value(
+                QFileInfo(fileToUpload._file).path(), std::numeric_limits<qint64>::max());
+    if (fileToUpload._size > quotaGuess) {
+        // Necessary for blacklisting logic
+        item->_httpErrorCode = 507;
+        emit propagator()->insufficientRemoteStorage();
+        done(item, SyncFileItem::DetailError, tr("Upload of %1 exceeds the quota for the folder").arg(Utility::octetsToString(fileToUpload._size)));
+        return;
+    }
+
+    qCDebug(lcBulkPropagatorJob) << "Running the compute checksum";
```

Comment:
Do you think that log statement is important for troubleshooting? Otherwise, I would tend to say better remove it because it might get executed a lot if people have a lot of files.